packages/doenetml-worker-rust/lib-doenetml-core/src/core/document_model/dependency_creation.rs

```diff
@@ -88,14 +88,39 @@ impl DocumentModel {
         };
 
         // Resolve a `PropComponent` to a component index.
-        let resolve_prop_component = |prop_component: &PropSource| match prop_component {
-            PropSource::Me => prop_pointer.component_idx,
-            PropSource::Parent => self
-                .document_structure
-                .borrow()
-                .get_true_component_parent(prop_pointer.component_idx)
-                .unwrap(),
-            PropSource::ByIdx(component_idx) => *component_idx,
+        let resolve_prop_component = |prop_component: &PropSource| {
+            Ok(match prop_component {
+                PropSource::Me => prop_pointer.component_idx,
+                PropSource::Parent => self
+                    .document_structure
+                    .borrow()
+                    .get_true_component_parent(prop_pointer.component_idx)
+                    .unwrap(),
+                PropSource::ByIdx(component_idx) => *component_idx,
+                PropSource::StaticComponentRef(local_prop_idx) => {
+                    // This is the prop that contains the ref.
+                    let prop_node = self.prop_pointer_to_prop_node(PropPointer {
+                        component_idx: prop_pointer.component_idx,
+                        local_prop_idx: *local_prop_idx,
+                    });
+                    // We could encounter a circularly-recursive call here if there was a programmer error.
+                    // If all goes well, `prop` will be a non-null ComponentRef.
+                    let prop = self.get_prop_untracked(prop_node, query_node);
```

Comment:
I'm really worried about calling `get_prop_untracked`, which will call `resolve_prop` while we are in the middle of `add_data_query` which is being called by `resolve_prop`. This makes it hard to reason what the code is doing and troubleshoot errors.

For example, what happens, if while you are looping over the data queries in line 45 of `prop_calculation.rs` and inside a call to `add_data_query`, the call to `get_prop_unchecked` ends up resolving one of the data queries further in the loop of line 45? Then, one will end up calling `add_data_query` twice on that data query, which will mess up the dependencies.

I'm not sure that my scenario is actually possible. But, this pattern opens up a lot more possibilities of code paths while we're creating the dependency graph that I feel could bite us hard as the code gets more complicated.

I'll keep thinking about this to see if there is a better pattern to get this result.

Reply:
What about this algorithm?

When a data query needs the value of a prop in order to create its dependency structure, then `add_data_query` checks if it is resolved. If it is resolved, then it gets the value as usual and proceeds.

However, if `add_data_query` needs the value of an unresolved `dependent_prop`, it aborts and reports that it needs  `dependent_prop` resolved. Then, `resolve_prop` adds the failed prop back to `resolve_stack` and subsequently adds `dependent_prop` to `resolve_stack` so that  `dependent_prop` will be resolved before the original prop.

The only wrinkle is that, the second time the original prop is popped off the stack, it may have already run `add_data_query` on some of its queries. So, we'd have to include bookkeeping about whether or not a data query has been added. (This bookkeeping might be necessary anyway when we have to recalculate queries in response to components to being dynamically added.)